core/runtime/src/fetch/response.rs

```diff
@@ -0,0 +1,236 @@
+pub enum ResponseType {
+    /// This applies in any of the following cases:
+    ///
+    /// The request is same-origin.
+    /// The requested URL's scheme is `data:`.
+    /// The request's mode is `navigate` or `websocket`.
+    ///
+    /// With this type, all response headers are exposed except Set-Cookie.
+    Basic,
+
+    /// The request was cross-origin and was successfully processed using CORS. With this
+    /// type, only CORS-safelisted response headers are exposed.
+    Cors,
+
+    /// A network error occurred. The status property is set to 0, `body` is null, headers
+    /// are empty and immutable.
+    Error,
+
+    /// A response to a cross-origin request whose mode was set to no-cors. The status
+    /// property is set to 0, `body` is null, headers are empty and immutable.
+    Opaque,
+
+    /// A response to a request whose redirect option was set to manual and which was
+    /// redirected by the server. The status property is set to 0, `body` is null, headers
+    /// are empty and immutable.
+    OpaqueRedirect,
+}
+
+impl ResponseType {
+    /// Return the JavaScript String representing this response type.
+    #[must_use]
+    pub fn to_string(self) -> JsString {
+        match self {
+            ResponseType::Basic => js_string!("basic"),
+            ResponseType::Cors => js_string!("cors"),
+            ResponseType::Error => js_string!("error"),
+            ResponseType::Opaque => js_string!("opaque"),
+            ResponseType::OpaqueRedirect => js_string!("opaqueredirect"),
+        }
+    }
+}
+
+impl TryFromJs for ResponseType {
+    fn try_from_js(value: &JsValue, context: &mut Context) -> JsResult<Self> {
+        let value_str = value.to_string(context)?;
+        if value_str == js_str!("basic") {
+            Ok(ResponseType::Basic)
+        } else if value_str == js_str!("cors") {
+            Ok(ResponseType::Cors)
+        } else if value_str == js_str!("error") {
+            Ok(ResponseType::Error)
+        } else if value_str == js_str!("opaque") {
+            Ok(ResponseType::Opaque)
+        } else if value_str == js_str!("opaqueredirect") {
+            Ok(ResponseType::OpaqueRedirect)
+        } else {
+            Err(js_error!(TypeError: "Invalid response type value"))
+        }
+    }
+}
+
+impl TryIntoJs for ResponseType {
+    fn try_into_js(&self, _: &mut Context) -> JsResult<JsValue> {
+        Ok(self.to_string().into())
+    }
+}
+
+/// The `Response` interface of the Fetch API represents the response to a request.
+//
+// You can create a new Response object using the `Response` constructor, but you
+// are more likely to encounter a `Response` object being returned as the result of
+// another API operation.
+#[derive(Clone, Debug, Trace, Finalize, JsData)]
+pub struct JsResponse {
+    url: JsString,
+
+    #[unsafe_ignore_trace]
+    r#type: ResponseType,
+
+    #[unsafe_ignore_trace]
+    status: Option<StatusCode>,
+
+    headers: JsHeaders,
+
+    body: Gc<Vec<u8>>,
```

Comment:
Since this doesn't have any `Gc` pointers in it, it would be better to use `Rc` instead.